docs/tutorial/source-plugin-tutorial.md

```diff
@@ -0,0 +1,984 @@
+
+```javascript:title=example-site/gatsby-config.js
+module.exports = {
+  plugins: [require.resolve(`../source-plugin`)],
+}
+```
+
+_You can include the plugin by using solely its name if you are using [npm link or yarn workspaces](/docs/creating-a-local-plugin/#using-npm-link-or-yarn-link) or place your `source-plugin` in [`example-site/plugins`](/docs/creating-a-local-plugin/) instead of being in a folder a step above and using `require.resolve`._
+
+You can now navigate into the `example-site` folder and run `gatsby develop`. You should see a line in the output in the terminal that shows your plugin loaded:
+
+```text
+$ gatsby develop
+success open and validate gatsby-configs - 0.033s
+success load plugins - 0.074s
+Loaded gatsby-starter-plugin // highlight-line
+success onPreInit - 0.016s
+...
+```
+
+If you open your `gatsby-node.js` file you will see the `console.log` that is resulting in the output in the terminal.
+
+```javascript:title=source-plugin/gatsby-node.js
+exports.onPreInit = () => console.log("Loaded gatsby-starter-plugin")
+```
+
+### Source data and create nodes
+
+Data is sourced in the `gatsby-node.js` file of source plugins or Gatsby sites. Specifically, it's done by calling a Gatsby function called `createNode` inside of the `sourceNodes` API in the `gatsby-node.js` file.
+
+#### Create nodes inside of `sourceNodes` with the `createNode` function
+
+Open up the `gatsby-node.js` file in the `source-plugin` and add the following code to create nodes from a hardcoded array of data :
+
+```javascript:title=source-plugin/gatsby-node.js
+// constants for your GraphQL Post and Author types
+const POST_NODE_TYPE = `Post`
+
+exports.sourceNodes = async ({
+  actions,
+  createContentDigest,
+  createNodeId,
+  getNodesByType,
+}) => {
+  const { createNode } = actions
+
+  const data = {
+    posts: [
+      { id: 1, description: `Hello world!` },
+      { id: 2, description: `Second post!` },
+    ],
+  }
+
+  // loop through data and create Gatsby nodes
+  data.posts.forEach(post =>
+    createNode({
+      ...post,
+      id: createNodeId(`${POST_NODE_TYPE}-${post.id}`),
+      parent: null,
+      children: [],
+      internal: {
+        type: POST_NODE_TYPE,
+        content: JSON.stringify(post),
+        contentDigest: createContentDigest(post),
+      },
+    })
+  )
+
+  return
+}
+```
+
+This code creates Gatsby node's that are queryable in a site. To the following steps break down what is happening in the code:
+
+- you implemented Gatsby's [`sourceNodes` API](/docs/node-apis/#sourceNodes) which Gatsby will run as part of its bootstrap process, and pulled out some Gatsby helpers (like `createContentDigest` and `createNodeId`) to facilitate creating nodes
```

Comment:
We should mention here that createContentDigest is important for tracking of dirty nodes and it should be included with the whole content of the item (`post` in this case)

Reply:
Agreed, some kind of link or explanation for what that is would be good.

Reply:
Good though! 

```suggestion
- you implemented Gatsby's [`sourceNodes` API](/docs/node-apis/#sourceNodes) which Gatsby will run as part of its bootstrap process, and pulled out some Gatsby helpers (like `createContentDigest` and `createNodeId`) to facilitate creating nodes
- you provided the required fields for the node like creating a node ID and a content digest (which Gatsby uses to track dirty nodes—or nodes that have changed), the content digest should include the whole content of the item (`post` in this case)
```